packages/web3/src/contract/script-simulator.test.ts

```diff
@@ -0,0 +1,292 @@
+    this.stack.push(val)
+  }
+
+  pop(): SimulatorVal {
+    const result = this.stack.pop()
+    if (result === undefined) {
+      throw new Error('Stack is empty')
+    }
+    return result
+  }
+
+  popBool(): any {
+    const result = this.pop()
+    if (result.kind !== 'Bool' && result.kind !== 'Symbol-Bool') {
+      throw new Error('Expected a Bool value on the stack')
+    }
+    return result
+  }
+}
+
+class LocalVariables {
+  private locals: any[] = []
+
+  get(index: number): any {
+    const result = this.locals[index]
+    if (result === undefined) {
+      throw new Error(`Local variable at index ${index} is not set`)
+    }
+    return result
+  }
+
+  set(index: number, val: any): void {
+    this.locals[index] = val
+  }
+
+  getBool(index: number): any {
+    const result = this.get(index)
+    if (result.kind !== 'Bool' && result.kind !== 'Symbol-Bool') {
+      throw new Error(`Local variable at index ${index} is not a Bool`)
+    }
+    return result
+  }
+}
+
+class ApprovedAccumulator {
+  private approvedTokens: { id: string; amount: bigint | 'unknown' }[] | 'unknown' = []
+
+  constructor() {
+    this.reset()
+  }
+
+  reset(): void {
+    this.approvedTokens = [{ id: ALPH_TOKEN_ID, amount: 0n }]
+  }
+
+  setUnknown(): void {
+    this.approvedTokens = 'unknown'
+  }
+
+  getApprovedAttoAlph(): bigint | 'unknown' | undefined {
+    if (this.approvedTokens === 'unknown') {
+      return 'unknown'
+    }
+    const approvedAttoAlph = this.approvedTokens[0].amount
+    return approvedAttoAlph === 0n ? undefined : approvedAttoAlph
+  }
+
+  getApprovedTokens(): any[] | 'unknown' | undefined {
+    if (this.approvedTokens === 'unknown') {
+      return 'unknown'
+    }
+    const tokens = this.approvedTokens.slice(1)
+    return tokens.length === 0 ? undefined : tokens
+  }
+
+  addApprovedAttoAlph(amount: any): void {
+    if (this.approvedTokens === 'unknown') {
+      return
+    }
+
+    switch (amount.kind) {
+      case 'U256':
+        this.approvedTokens[0].amount += amount.value
+        break
+
+      case 'Symbol-U256':
+        this.setUnknown()
+        break
+
+      case 'undefined':
+      case null:
+        this.setUnknown()
+        break
+
+      default:
+        this.setUnknown()
+        break
+    }
+  }
+
+  addApprovedToken(tokenId: any, amount: any): void {
+    if (this.approvedTokens === 'unknown') {
+      return
+    }
+
+    if (tokenId.kind !== 'ByteVec') {
+      this.setUnknown()
+      return
+    }
+
+    switch (amount.kind) {
+      case 'U256':
+        const id = binToHex(tokenId.value)
+        const existing = this.approvedTokens.find((t) => t.id === id)
+
+        if (existing) {
+          existing.amount += amount.value
+        } else {
+          this.approvedTokens.push({ id, amount: amount.value })
+        }
+        break
+
+      case 'Symbol-U256':
+        this.setUnknown()
+        break
+
+      default:
+        this.setUnknown()
+        break
+    }
+  }
```

Comment:
This whole section imho introduces extra complexity and a level of indirection, if we look at what you try to mock, the `addApprovedAttoAlph` just calls `addApprovedToken`  : 
```
  addApprovedAttoAlph(amount: ValU256 | SymbolU256): void {
    this.addApprovedToken({ kind: 'ByteVec', value: hexToBinUnsafe(ALPH_TOKEN_ID) }, amount)
  }
  addApprovedToken(tokenId: ValByteVec | SymbolByteVec, amount: ValU256 | SymbolU256): void {
```
Can it be mimicked ?